### ISSUE TITLE
[discussion] Should Traverson set the Accept and Content-type header automatically?

### ISSUE THREAD

**Issue by jkschneider** [basti1302/traverson#37]
This code will fail if the root API endpoint (http://localhost:8080/api) only responds to `application/hal+json`:

``` javascript
var traverson = require('traverson');
var JsonHalAdapter = require('traverson-hal');

traverson.registerMediaType(JsonHalAdapter.mediaType, JsonHalAdapter);

traverson.from('http://localhost:8080/api')
  .jsonHal()
  .follow('mylink')
  .getResource(function(e, d) {
      // ....
  });
```

Manually setting a request option fixes it, but it seems like this should be the default:

``` javascript
traverson.from('http://localhost:8080/api')
  .jsonHal()
  .withRequestOptions({ headers: { accept: 'application/hal+json' }})
```

If this seems reasonable, just say so and I'll give you a PR.

**Comment by basti1302** [comment]
I though about setting Accept/Content-Type headers automatically but decided against it until now. Especially when it comes to using the .json()` variant because the server might just use some `application/foo+json` content type and if we set accept to `application/json` automatically this might trip up content negotiation.

That's why I decided against it and force the user to always explicitly set these headers.

Now with traverson-hal, which really only deals with `application/hal+json` this might be a different story. But even the HAL reference server does not deal with `Accept: application/hal+json` correctly (see https://github.com/mikekelly/hal-browser/issues/35).

As I'm quite ambivalent on the issue I'll leave this open for discussion. If enough people think that default headers should be set and have good reasons, I might be swayed.

**Comment by basti1302** [comment]
I took the liberty to change the title of the issue, hope you don't mind.

**Comment by hnilsen** [comment]
Have you given this any more thought? I agree with @jkschneider that I was expecting the accept-header to be set when it was set when the mediaType is set.

**Comment by travi** [comment]
Keep in mind that it is [very common](https://dzone.com/articles/your-api-versioning-wrong) to specify version in the accept header. For example, my company uses `application/vnd.<company-name>.v<version-number>.hal+json` for the accept header. We also expect this header as the content type for requests that include a body.

While it may make sense for this to be a default for simple hal APIs, I don't think it would end up being correct for a large number of more mature APIs. I think I agree with @basti1302 that having a default here might end up causing more confusion than making it clear that it needs to be set explicitly.

**Comment by hnilsen** [comment]
I know that. But application/hal+json is the baseline, and one should not expect the content type to magically understand your own types and versioning. I don't understand that argument.

Accept header with "application/hal+json" should be the default when using this command:
`traverson.registerMediaType(JsonHalAdapter.mediaType, JsonHalAdapter);`

and running the command:
`.jsonHal()`

If you need your own accept header, you should specify that explicitly, that's a given.

**Comment by jkschneider** [comment]
+1 on @hnilsen's opinion. I don't understand the argument that having an overridable default is more confusing than no default at all if your organization uses a more specialized content type.

**Comment by basti1302** [comment]
Okay, I would go with having the next major version of traverson and traverson-hal setting default accept and Content-type headers. The API for media type plug-ins probably has to be extended for this as a media type plug-in needs to decide about the values for these headers. Also, this needs to be added to the docs.

Is any of you interested in providing a PR for this?

**Comment by basti1302** [comment]
After close to two years, this has now landed in 6.0.0.
